File: Py/cs50/credit.c
#include <stdio.h>
#include <cs50.h>
#include <string.h>
#include <ctype.h>

bool digits(char *ccString);
    
int main(void)
{
    char *ccString = get_string("Number: ");
    int ccLen = strlen(ccString);
    int l = 0;
    while (!digits(ccString) || ccLen == 0)
    {
        ccString = get_string("Number: ");
        ccLen = strlen(ccString);
    }
    int ccNum[ccLen];
    for (int i = 0; i < ccLen; i++)
    {
        ccNum[i] = ccString[i] - 48;
    }
    int ccLuhn[ccLen];;
    int sumLuhn = 0;
    for (int j = ccLen - 2; j >= 0; j -= 2)
    {
        ccLuhn[j] = 2 * ccNum[j] / 10 + 2 * ccNum[j] % 10;
        sumLuhn += ccLuhn[j];
        
    }
    for (int k = ccLen - 1; k >= 0; k -= 2)
    {
        sumLuhn += ccNum[k];
    }
    if (sumLuhn % 10 != 0 || ccLen < 13)
    {
        printf("INVALID\n");
        return 0;
    }
    if (ccLen == 15 && ccNum[0] == 3 && (ccNum[1] == 4 || ccNum[1] == 7))
    {
        printf("AMEX\n");
    }
    else if (ccLen == 16 && ccNum[0] == 5 && (ccNum[1] >= 1 && ccNum[1] <= 5))
    {
        printf("MASTERCARD\n");
    }
    else if ((ccLen >= 13 && ccLen <= 16) && ccNum[0] == 4)
    {
        printf("VISA\n");
    }
    else
    {
        printf("INVALID\n");
    }
}

bool digits(char *ccString)
{
    for (int w = 0; w < strlen(ccString); w++)
    {
        if (!isdigit(ccString[w]))
        {
            return false;
        }
    }
    return true;
}

File: Py/cs50/pset6/credit.py
from cs50 import get_string

# get # from user, check if valid input


def getNum():
    while True:
        ccNum = get_string("CC Number: ")
        if ccNum.isdigit():
            return ccNum

# once valid, check if LUHNS is good


def luhn(ccNum):
    ccLuhn = ccNum[len(ccNum)-2::-2]
    ccLuhn2 = ccNum[len(ccNum)-1::-2]
    a = 0
    b = 0
    for c in ccLuhn:
        a += 2*int(c) // 10 + 2*int(c) % 10
    for c in ccLuhn2:
        b += int(c)
    luhns = (a + b) % 10
    if (luhns == 0):
        return True
    return False

# once luhns is good, check other things


def cardType(ccNum):
    ccLen = len(ccNum)
    first = int(ccNum[0])
    second = int(ccNum[1])
    print(f"{ccNum[0]}{ccLen}")
    if ccLen == 15 and first == 3 and (second == 4 or second and 7):
        print("AMEX")
    elif ccLen == 16 and first == 5 and (second >= 1 and second <= 5):
        print("MASTERCARD")
    elif (ccLen >= 13 and ccLen <= 16) and first == 4:
        print("VISA")
    else:
        print("INVALID")
    return True


# driver
ccNum = getNum();
if luhn(ccNum):
    cardType(ccNum)
else:
    print("INVALID")
